File: recursion/python/solutions/merge.py
"""Solution for merge."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def merge(lst1, lst2):
    """
    Merge two SORTED lists into one sorted list.

    merge(1 -> 3 -> 5, 2 -> 4 -> 6) => 1 -> 2 -> 3 -> 4 -> 5 -> 6
    merge(1 -> 2, ()) => 1 -> 2
    """
    if is_empty(lst1):
        return lst2

    if is_empty(lst2):
        return lst1

    first1, rest1 = unprepend(lst1)
    first2, rest2 = unprepend(lst2)

    if first1 <= first2:
        return prepend(first1, merge(rest1, lst2))
    else:
        return prepend(first2, merge(lst1, rest2))


if __name__ == '__main__':
    # Test cases
    assert to_python_list(merge(Node.from_list([1, 3, 5]), Node.from_list([2, 4, 6]))) == [1, 2, 3, 4, 5, 6]
    assert to_python_list(merge(Node.from_list([1, 2]), EMPTY_LIST)) == [1, 2]
    assert to_python_list(merge(EMPTY_LIST, Node.from_list([1, 2]))) == [1, 2]
    assert to_python_list(merge(Node.from_list([1, 2, 3]), Node.from_list([1, 2, 3]))) == [1, 1, 2, 2, 3, 3]

    print("All tests passed!")

File: recursion/javascript/solutions/merge.js
/**
 * Solution for merge.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function merge(list1, list2) {
  if (isEmpty(list1)) {
    return list2;
  }

  if (isEmpty(list2)) {
    return list1;
  }

  let [first1, rest1] = unprepend(list1);
  let [first2, rest2] = unprepend(list2);

  if (first1 <= first2) {
    return prepend(first1, merge(rest1, list2));
  }
  return prepend(first2, merge(list1, rest2));
}

if (require.main === module) {
  console.assert(JSON.stringify(toArray(merge(Node.fromArray([1, 3, 5]), Node.fromArray([2, 4, 6])))) === '[1,2,3,4,5,6]');
  console.assert(JSON.stringify(toArray(merge(Node.fromArray([1, 2]), EMPTY_LIST))) === '[1,2]');
  console.assert(JSON.stringify(toArray(merge(EMPTY_LIST, Node.fromArray([1, 2])))) === '[1,2]');
  console.assert(JSON.stringify(toArray(merge(Node.fromArray([1, 1]), Node.fromArray([1, 1])))) === '[1,1,1,1]');

  console.log('All tests passed!');
}

module.exports = { merge };
